### PR TITLE
Exposed API for xeve input arguments handling and redefined structure describing command line option

### PR DESCRIPTION
**Pull request by dkozinski**

### CODE REVIEW

**Comment by dkozinski** [comment]
Xeve ARGS API has been exposed to meet the expectations of the FFmpeg community.
Currently, our FFmpeg xeve wrapper implementation enables the user to use only a very limited set of options in order to drive xeve codec parameters.
Some time ago, we made an attempt to map and handle more options available for the xeve reference application to be available also for FFmpeg xeve wrapper.
Unfortunately, FFmpeg community rejected our changes and suggested that we should do this a different way,  as other codec providers do ( see hevc, h264).
All options available for the xeve reference app should be handled using the xeve library. We should be able to pass them to the FFmpeg xeve wrapper as a key-value list, and the wrapper in turn should delegate all the parsing them to the xeve library.

File: inc/xeve_args.h [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+ * @param args
+ * @param argc number of command line argumnets (strings in argv) passed to main()
+ * @param argv table containing command line arguments (strings) pased to main
+ * @param errstr
+ * @return int
+ */
+int XEVE_EXPORT xeve_args_parse(ARGS_PARSER * args, int argc, const char* argv[], char ** errstr);
+
+// zwraca kod bledu jesli opcja nie jest ustawiona lub niedostepna (nie ma takiej opcji)
+// Sa opcje ktore poiadaja wartosc numeryczna. w gesti uzytkownika pozostaje ewentualna konwersja do int
```
Comment (dariusz-f) [inline inc/xeve_args.h:80]: Remove Polish comment

File: inc/xeve_args.h [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+ * @retval XEVE error code if Error
+ */
+int XEVE_EXPORT xeve_args_init(ARGS_PARSER * args, XEVE_PARAM* param);
+
+/**
+ * @brief Parse application input arguments
+ *
+ * @param args
+ * @param argc number of command line argumnets (strings in argv) passed to main()
+ * @param argv table containing command line arguments (strings) pased to main
```
Comment (dariusz-f) [inline inc/xeve_args.h:73]: add here () to main or remove in line above

File: inc/xeve_args.h [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+ * the next key value after the one passed as an argument.
+ * If the passed key value is NULL returns the first key value from the options container.
+ *
+ * @param args
+ * @param key string containg key value or NULL
+ *
+ * @retval NULL if there is no next key in the options container
+ * @retval key value Ok
+ */
+char* XEVE_EXPORT xeve_args_next_key(ARGS_PARSER * args, char* key);
```
Comment (dariusz-f) [inline inc/xeve_args.h:155]: Remove this functions

File: src_base/xeve_args.c [hunk 4]
```diff
@@ -0,0 +1,2313 @@
...
+        StrOption string;
+    } default_value;                        /* default value */
+
+    ValuesRageType values_rage_type;        /* range or list */
+    union {
+        Range range;
+        List list;
+    } allowed_values;                       /* allowed values */
+
+    int is_mandatory;                       /* flag: 0 - option is optional; 1 - option is mandatory */
```
Comment (dariusz-f) [inline src_base/xeve_args.c:109]: is this still required?

File: src_base/xeve_args.c [hunk 5]
```diff
@@ -0,0 +1,2313 @@
...
+                                            .allowed_values =  { .list = { .size = 0, .strings =  {{NULL,NULL}} } },
+                                            .is_mandatory = 0, .is_set = 0,
+                                            .desc = "file name of reconstructed video",
+                                            .opt_storage = NULL },
+
+        // width
+        { .key = 'w', .key_long = "width", .value_type = VAL_TYPE_INT,
+                                            .has_default = 0, .default_value = { .integer = { 0, NULL } },
+                                            .values_rage_type = AVT_MIN_MAX_RANGE,
+                                            .allowed_values =  { .range = {128,3840} }, // SQCIF	128 × 96 / 4K 3840 x 2160 / 8K 7680 x 4320
```
Comment (dariusz-f) [inline src_base/xeve_args.c:157]: what with 8k?

File: src_base/xeve_args.c [hunk 6]
```diff
@@ -0,0 +1,2313 @@
...
+    if( !args )
+        return NULL;
+
+    if(!args->num_option) return NULL;
+
+    opt = args->opts;
+    return opt->key_long;
+}
+
+// int nums_done(ARGS_PARSER * args, char* key);        /* returns 1 if key is past end */
```
Comment (dariusz-f) [inline src_base/xeve_args.c:2079]: remove

File: src_base/xeve_args.c [hunk 7]
```diff
@@ -0,0 +1,2313 @@
...
+        oidx = args_search_long_key(ops, tag);
+        if (oidx < 0) continue;
+
+        if (ops[oidx].opt_storage == NULL)
+        {
+            return -1;
+        }
+
+        //if (ARGS_GET_IS_OPT_TYPE_PPT(ops[oidx].value_type) == is_type_ppt)
+        //{
```
Comment (dariusz-f) [inline src_base/xeve_args.c:2203]: remove

File: src_base/xeve_args.c [hunk 8]
```diff
@@ -0,0 +1,2313 @@
...
+            if (ops[oidx].value_type == VAL_TYPE_INT || ops[oidx].value_type == VAL_TYPE_STRING)
+            {
+                if (args_read_value(ops + oidx, val)) continue;
+            }
+            else
+            {
+                *((int*)ops[oidx].opt_storage) = 1;
+            }
+            ops[oidx].is_set = 1;
+        //}
```
Comment (dariusz-f) [inline src_base/xeve_args.c:2213]: remove

File: src_base/xeve_args.c [hunk 9]
```diff
@@ -0,0 +1,2313 @@
...
+
+
+    } else {
+        next_idx = 0;
+    }
+
+    opt = args->opts + next_idx;
+
+    return opt->key_long;
+}
```
Comment (dariusz-f) [inline src_base/xeve_args.c:2108]: is this functions is still required?

**Review by dariusz-f — CHANGES_REQUESTED** [review]

File: inc/xeve_args.h (hunk 3 above)
Comment (dkozinski) [reply inc/xeve_args.h:155]: xeve_args_get_option_description(ARGS_PARSER * args, char* keyl, char ** help) function needs the key value to provide a description for a given option related to that key. And to get the subsequent keys the 
 xeve_args_next_key() function is needed. It makes it able to get all subsequent keys for all options.

Comment (dkozinski) [reply inc/xeve_args.h:155]: As it comes to xeve_args_check_mandatory() it is still used by xeve_app to check whether all mandatory options are set.

File: app/xeve_app.c [hunk 10]
```diff
@@ -192,29 +278,84 @@ static void print_stat_init(ARGS_PARSER * args)
...
     }
-    if(strlen(args->fname_rec) > 0)
+    if (fname_rec_str && strlen(fname_rec_str) > 0)
     {
-        logv2("Output YUV file         : %s \n", args->fname_rec);
+        logv2("Output YUV file         : %s \n", fname_rec_str);
     }
 
-    if (strlen(args->fname_cfg) > 0)
+    if (fname_cfg_str && fname_cfg_str && strlen(fname_cfg_str) > 0)
```
Comment (washish) [inline app/xeve_app.c:350]: fname_cfg_str is used twice in this conditional block

File: src_base/xeve_args.c [hunk 11]
```diff
@@ -0,0 +1,2313 @@
...
+    }
+    return ret;
+}
+
+int xeve_args_check_mandatory(ARGS_PARSER * args, char ** err_arg)
+{
+    ARGS_OPT * o = args->opts;
+
+    // do some checks
+    if(!args)
```
Comment (washish) [inline src_base/xeve_args.c:1818]: This check should be before the line ARGS_OPT * o = args->opts;

File: src_base/xeve_args.c [hunk 12]
```diff
@@ -0,0 +1,2313 @@
...
+    memcpy(opts, args_opt_table, sizeof(args_opt_table));
+    args->opts = opts;
+    args->num_option = ((int)(sizeof(args_opt_table)/sizeof(args_opt_table[0]))-1);
+    args->initialized = 0;
+
+    return args;
+
+ERR:
+    if(opts) {
+        free(opts);
```
Comment (washish) [inline src_base/xeve_args.c:1419]: This line might be unreachable as opts value will be NULL in both the goto flow.

**Comment by mpeg5** [comment]
This changes need to be checked carefully. I will check x264 case.

File: src_base/xeve_args.c [hunk 13]
```diff
@@ -0,0 +1,2313 @@
...
+    if(argv[0] == '-' && (argv[1] < '0' || argv[1] > '9')) return -1;
+
+    switch(ops->value_type)
+    {
+        case VAL_TYPE_INT:
+            *((int*)ops->opt_storage) = atoi(argv);
+            break;
+
+        case VAL_TYPE_STRING:
+            strcpy((char*)ops->opt_storage, argv);
```
Comment (washish) [inline src_base/xeve_args.c:2166]: This function is unsafe, use strncpy instead.

File: src_base/xeve_args.c (hunk 4 above)
Comment (dkozinski) [reply src_base/xeve_args.c:109]: Yes it is

File: src_base/xeve_args.c (hunk 9 above)
Comment (dkozinski) [reply src_base/xeve_args.c:2108]: Yes, it is. xeve_args_get_option_description(ARGS_PARSER * args, char* keyl, char ** help) function needs the key value to provide a description for a given option related to that key. And to get the subsequent keys the
xeve_args_next_key() function is needed. It makes it able to get all subsequent keys for all options.

File: src_base/xeve_args.c (hunk 11 above)
Comment (dkozinski) [reply src_base/xeve_args.c:1814]: DONE

File: src_base/xeve_args.c (hunk 13 above)
Comment (dkozinski) [reply src_base/xeve_args.c]: > [washish](/washish)

x264.h
int x264_param_parse( x264_param_t *, const char *name, const char *value );

File: src_base/xeve_args.c [hunk 14]
```diff
@@ -0,0 +1,2271 @@
...
+            len = strlen(buf);
+            *val = (char*)malloc(len+1);
+            strncpy(*val, buf, len+1);
+            // *val[len-1] = '\0';
+
+            break;
+        case VAL_TYPE_STRING:
+
+            str = (char*)(args->opts[idx].opt_storage);
+            if(!str) return XEVE_ERR;
```
Comment (washish) [inline src_base/xeve_args.c:1877]: Line 1855 already does the check for NULL args->opts[idx].opt_storage, so this if block might be unreachable. please check

**Comment by kpchoi** [comment]
@dkozinski @washish 

I tried to find similar code in libx264.c of ffmpeg.
I could not find including "x264cli.h"
Can you see the similar code including?

**Comment by kpchoi** [comment]
> @dkozinski @washish
> 
> I tried to find similar code in libx264.c of ffmpeg. I could not find including "x264cli.h" Can you see the similar code including?

Do we need to add API like "x264_param_parse(&x4->params, opt, param)" in xeve.h?
And then, it would be fine, whole exposing of argument parser would not be good.

**Comment by dkozinski** [comment]
@kpchoi 
> > @dkozinski @washish
> > I tried to find similar code in libx264.c of ffmpeg. I could not find including "x264cli.h" Can you see the similar code including?
> 
> Do we need to add API like "x264_param_parse(&x4->params, opt, param)" in xeve.h? And then, it would be fine, but whole exporting of the argument parser would not be good.

Yes. It's exactly what we need.

The reason why we decided to expose the entire argument parser API was the extension of the API with functions that make it easier for the user to create applications using the xeve library.

But in fact at the moment xeve_param_parse function is basically all we need for xeve ffmpeg implementation.

So what we can do right now (in this PR) is get rid of the public xeve_args.h header and the xeve_args.c implementation from the xeve library and expose (as you have suggested) xeve_param_parse (XEVE_PARAM * param, const char * name, const char * value) function in xeve.h, as well as provide implementation of the function xeve_param_parse in the xeve_param_parse.c file.

As for the files xeve_args.h and xeve_args.c, they could possibly be used in the implementation of the reference application.

**Comment by washish** [comment]
> So what we can do right n



> @dkozinski @washish
> 
> I tried to find similar code in libx264.c of ffmpeg. I could not find including "x264cli.h" Can you see the similar code including?
@kpchoi 
Line no. 2393-97 have some explanation about the way params are handled for x264.
https://github.com/FFmpeg/FFmpeg/blob/bd7d69fe354cf5b3d6247f049af44bfa56f0531a/doc/encoders.texi#L2393

**Comment by kpchoi** [comment]
> > So what we can do right n
> 
> > @dkozinski @washish
> > I tried to find similar code in libx264.c of ffmpeg. I could not find including "x264cli.h" Can you see the similar code including?
> > @kpchoi
> > Line no. 2393-97 have some explanation about the way params are handled for x264.
> > https://github.com/FFmpeg/FFmpeg/blob/bd7d69fe354cf5b3d6247f049af44bfa56f0531a/doc/encoders.texi#L2393

When I saw the source code in detail, I found that x264cli.h is not included by ffmpeg.
ffmpeg is using just x264_param_parse() only.

**Comment by kpchoi** [comment]
> @kpchoi
> 
> > > @dkozinski @washish
> > > I tried to find similar code in libx264.c of ffmpeg. I could not find including "x264cli.h" Can you see the similar code including?
> > 
> > 
> > Do we need to add API like "x264_param_parse(&x4->params, opt, param)" in xeve.h? And then, it would be fine, but whole exporting of the argument parser would not be good.
> 
> Yes. It's exactly what we need.
> 
> The reason why we decided to expose the entire argument parser API was the extension of the API with functions that make it easier for the user to create applications using the xeve library.
> 
> But in fact at the moment xeve_param_parse function is basically all we need for xeve ffmpeg implementation.
> 
> So what we can do right now (in this PR) is get rid of the public xeve_args.h header and the xeve_args.c implementation from the xeve library and expose (as you have suggested) xeve_param_parse (XEVE_PARAM * param, const char * name, const char * value) function in xeve.h, as well as provide implementation of the function xeve_param_parse in the xeve_param_parse.c file.
> 
> As for the files xeve_args.h and xeve_args.c, they could possibly be used in the implementation of the reference application.

I agree that we add xeve_param_parse() function as like x264 library, and it could be called by ffmpeg.
Can you handle this?

**Comment by dkozinski** [comment]
> 



> > @kpchoi
> > > > @dkozinski @washish
> > > > I tried to find similar code in libx264.c of ffmpeg. I could not find including "x264cli.h" Can you see the similar code including?
> > > 
> > > 
> > > Do we need to add API like "x264_param_parse(&x4->params, opt, param)" in xeve.h? And then, it would be fine, but whole exporting of the argument parser would not be good.
> > 
> > 
> > Yes. It's exactly what we need.
> > The reason why we decided to expose the entire argument parser API was the extension of the API with functions that make it easier for the user to create applications using the xeve library.
> > But in fact at the moment xeve_param_parse function is basically all we need for xeve ffmpeg implementation.
> > So what we can do right now (in this PR) is get rid of the public xeve_args.h header and the xeve_args.c implementation from the xeve library and expose (as you have suggested) xeve_param_parse (XEVE_PARAM * param, const char * name, const char * value) function in xeve.h, as well as provide implementation of the function xeve_param_parse in the xeve_param_parse.c file.
> > As for the files xeve_args.h and xeve_args.c, they could possibly be used in the implementation of the reference application.
> 
> I agree that we add xeve_param_parse() function as like x264 library, and it could be called by ffmpeg. Can you handle this?

Yes, we'll handle this.

**Comment by dkozinski** [comment]
Due to the fact that finally a lot of stuff from the current PR will be rejected, the implementation of xeve_param_parse() function has been moved to the new PR.

**Pull request closed without merging**